aconcli/service/aconclient_http.go

```diff
@@ -0,0 +1,465 @@
+	}
+	w.Close()
+
+	req, err := http.NewRequest(http.MethodPut, requestURL, body)
+	if err != nil {
+		return fmt.Errorf(clientMakeReqErrFmt, "AddBlob", err)
+	}
+	req.Header.Add("Content-Type", w.FormDataContentType())
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "AddBlob", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "AddBlob", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Finalize() error {
+	requestURL := c.makeURL(endpointFinalize)
+	req, err := http.NewRequest(http.MethodPost, requestURL, nil)
+	if err != nil {
+		return fmt.Errorf(clientMakeReqErrFmt, "Finalize", err)
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "Finalize", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "Finalize", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Start(imageId string, env []string) (uint32, error) {
+	requestURL := c.makeURL(endpointStart)
+	d := url.Values{}
+	d.Set(fieldImgeId, imageId)
+	for _, e := range env {
+		d.Add(fieldEnvs, e)
+	}
+
+	resp, err := c.client.PostForm(requestURL, d)
+	if err != nil {
+		return 0, fmt.Errorf(clientSendReqErrFmt, "Start", err)
+	}
+	content, err := processReponse(resp)
+	if err != nil {
+		return 0, fmt.Errorf(clientProcRespErrFmt, "Start", err)
+	}
+	r := StartResponse{}
+	if err = json.Unmarshal(content, &r); err != nil {
+		return 0, fmt.Errorf(clientUnmarshalErrFmt, "Start", err)
+	}
+	return r.ContainerId, nil
+}
+
+func (c *AconClientHttp) Kill(cid uint32, signum int32) error {
+	requestURL := c.makeURL(endpointKill)
+	d := url.Values{}
+	d.Set(fieldConId, strconv.FormatUint(uint64(cid), 10))
+	d.Set(fieldSignum, strconv.FormatInt(int64(signum), 10))
+
+	resp, err := c.client.PostForm(requestURL, d)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "Kill", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "Kill", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Restart(cid uint32, timeout uint64) error {
+	requestURL := c.makeURL(endpointRestart)
+	d := url.Values{}
+	d.Set(fieldConId, strconv.FormatUint(uint64(cid), 10))
+	d.Set(fieldTimeout, strconv.FormatUint(timeout, 10))
+
+	resp, err := c.client.PostForm(requestURL, d)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "Restart", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "Restart", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Invoke(cid uint32, invocation []string,
+	timeout uint64, env []string, datafile string, capture_size uint64) ([]byte, []byte, error) {
+	requestURL := c.makeURL(endpointExec)
+	body := &bytes.Buffer{}
+	w := multipart.NewWriter(body)
+	if datafile != "" {
+		if err := multipartFile(w, fieldStdin, datafile); err != nil {
```

Comment:
It seems stdin is the only file, while other parameters can be passed via URL. Why use HTTP form here? Please note the query parameters will be encoded in both URL and HTTP header so integrity is protected by TLS.